#'Plot read profiles for a particular feature.
#'
#'\code{plotFeature} plots the achieved performance for each feature/gene.
#'The resulting graph shows one bar per each feature/gene with heights
#'according to its attribute value. If complete is set as TRUE two bar plots
#'(feature and gene level) will be stored in the resulting ggplot object.
#'
#'@param object TargetExperiment object.
#'@param featureID Character indicating the ID of the feature.
#'@param SNPs Logical flag indicating if SNPs should be plotted.
#'@param xlab Character containing the axis x label.
#'@param title Character containing the plot title.
#'@param size Numeric indicating the size of line plots.
#'@param BPPARAM An optional BiocParallelParam instance defining the parallel
#'back-end to be used during evaluation.
#'
#'@return ggplot2 graphics.
#'
#'@include TargetExperiment-plotRegion.R
#'@exportMethod plotFeature
#'@docType methods
#'@name plotFeature
#'@rdname TargetExperiment-plotFeature
#'@aliases plotFeature-methods
#'@seealso \code{\link{plotRegion}}
#'@note see full example in \code{\link{TargetExperiment-class}}
#'@author Gabriela A. Merino \email{gmerino@@bdmg.com.ar}, Cristobal Fresno
#'\email{cfresno@@bdmg.com.ar} and Elmer A. Fernandez 
#'\email{efernandez@@bdmg.com.ar}
#'@examples
#'## loading TargetExperiment object
#'data(ampliPanel, package="TarSeqQC")
#'## Defining bam file, bed file and fasta file names and paths
#'setBamFile(ampliPanel)<-system.file("extdata", "mybam.bam", 
#'    package="TarSeqQC", mustWork=TRUE)
#'setFastaFile(ampliPanel)<-system.file("extdata", "myfasta.fa", 
#'    package="TarSeqQC", mustWork=TRUE)
#'
#'# Exploring the read count profile for a particular amplicon
#'g<-plotFeature(ampliPanel, featureID="AMPL20")
#'if(interactive()){
#'g
#'}
setGeneric(name="plotFeature", def=function(object,featureID, SNPs=TRUE, 
xlab="", title="",size=0.5, BPPARAM=bpparam()){
    standardGeneric("plotFeature")
})
#'
#'@name plotFeature
#'@rdname TargetExperiment-plotFeature
#'@import BiocParallel
#'@importMethodsFrom BiocParallel bpprogressbar<-
#'@aliases plotFeature,TargetExperiment-method
#'@inheritParams plotFeature
setMethod(f="plotFeature", signature=signature(object="TargetExperiment"),
definition=function(object, featureID, SNPs=TRUE,  xlab="", title=featureID, 
size=0.5, BPPARAM=bpparam()){
    bpprogressbar(BPPARAM)<-TRUE
    bed_file<-getBedFile(object)
    
    if(!(featureID %in% names(bed_file))) {
        stop(paste("The feature", featureID, "is not contained in the bed file",
        sep=" "))
    }
    
    bed_feature<-bed_file[names(bed_file) == featureID]
    if(length(bed_feature) >1){
        stop(paste("The bed file contains two regions with the same name: ",
        featureID, sep =""))
    }
    p<-plotRegion(object,region=c(start(bed_feature), end(bed_feature)), 
        seqname=runValue(seqnames(bed_feature)), SNPs,  xlab, title, size, 
        BPPARAM)

    return(p)

})

